projects/flow.groups-2D-random-walk/src/project.py

```diff
@@ -0,0 +1,211 @@
+
+
+@base
+@RandomWalkProject.pre.after(simulate)
+@RandomWalkProject.post(lambda job: "squared_distance" in job.data)
+@RandomWalkProject.operation
+def compute_squared_distance(job):
+    """Compute the squared distance for a random walk."""
+    with job.data:
+        positions = job.data.positions[:]
+    job.data.squared_distance = np.sum(positions * positions, axis=1)
+
+
+# This operation happens after computing the msd so it isn't in base.
+# Also we use pre as a filter for jobs ensuring this job only ever runs on the zeroth
+# replica for a given standard deviation.
+@RandomWalkProject.pre.true("msd_analyzed")
+@RandomWalkProject.pre(lambda job: job.sp.replica == 0)
+@RandomWalkProject.post.isfile("msd.png")
+@RandomWalkProject.operation
+def plot_msd(job):
+    """Plot the MSD for all std."""
+    with job.data:
+        msd = job.data.msd[:]
+    fig, ax = plt.subplots()
+    ax.plot(msd)
+    ax.set_title(f"MSD for {job.sp.standard_deviation} std")
+    ax.set_xlabel("x")
+    ax.set_ylabel("MSD")
+    # Only save figure to the first replica
+    fig.savefig(job.fn("msd.png"))
+    fig.close()
+
+
+# Aggregate operations on individual standard deviations across replicas
+
+# Create aggregator that combines all replicas with a single standard deviation
+std_aggregator = flow.aggregator.groupby("standard_deviation")
+
+
+# Define all aggregate groups
+agg_plot = RandomWalkProject.make_group(
+    "aggregate-plot", group_aggregator=std_aggregator
+)
+agg_analyze_and_plot = RandomWalkProject.make_group(
+    "post-processing", group_aggregator=std_aggregator
+)
+
+
+@agg_analyze_and_plot
+@std_aggregator
+@RandomWalkProject.pre(all_simulated)
+@RandomWalkProject.post.true("msd_analyzed")
+@RandomWalkProject.operation
+def compute_mean_squared_distance(*jobs):
+    """Compute and store the mean squared displacement for all std."""
+    squared_distance_iterator = generate_stores(jobs, "squared_distance")
+    squared_distance = next(squared_distance_iterator)
+    msd = np.copy(squared_distance)
+    for squared_distance in squared_distance_iterator:
+        msd += squared_distance
+    msd /= len(jobs)
+    # Store msd in only first replica (job.sp.replica == 1)
+    job_replica_zero = get_replica(jobs, replica=0)
+    job_replica_zero.data["msd"] = msd
+    for job in jobs:
+        job.doc.msd_analyzed = True
+
+
+# Since this uses a separate aggragator to restrict aggregates to the first 5 replicas,
+# we cannot assign this operation to either agg_plot or agg_analyze_and_plot
+@flow.aggregator.groupby("standard_deviation", select=lambda job: job.sp.replica <= 4)
+@RandomWalkProject.pre(all_simulated)
+@RandomWalkProject.post.true("plotted_walks")
+@RandomWalkProject.operation
+def plot_walk(*jobs):
+    """Plot the first 5 replicas random walks for each standard_deviation."""
+    fig, ax = plt.subplots()
+    for position, job in zip(generate_stores(jobs, "positions"), jobs):
```

Comment:
Should we add a "sort by" to the aggregator so that the labels/lines are in order? That could also eliminate the need for `get_replica(jobs, replica=0)` because it would be `jobs[0]` by definition.